crates/bevy_ecs/src/world/mod.rs

```diff
@@ -2796,4 +2987,63 @@ mod tests {
         let mut world = World::new();
         world.spawn(());
     }
+
+    #[test]
+    fn iterate_resources() {
+        let mut world = World::new();
+
+        world.insert_resource(TestResource(42));
+        world.insert_resource(TestResource2("Hello, world!".to_string()));
+
+        let mut iter = world.iter_resources();
+
+        let (info, ptr) = iter.next().unwrap();
+        assert_eq!(info.name(), std::any::type_name::<TestResource>());
+        // SAFETY: We know that the resource is of type `TestResource`
+        assert_eq!(unsafe { ptr.deref::<TestResource>().0 }, 42);
+
+        let (info, ptr) = iter.next().unwrap();
+        assert_eq!(info.name(), std::any::type_name::<TestResource2>());
+        assert_eq!(
+            // SAFETY: We know that the resource is of type `TestResource2`
+            unsafe { &ptr.deref::<TestResource2>().0 },
+            &"Hello, world!".to_string()
+        );
+
+        assert!(iter.next().is_none());
+    }
+
+    #[test]
+    fn iterate_resources_mut() {
```

Comment:
These need a test case to ensure that iteration doesn't panic if you remove a resource. Also for miri.